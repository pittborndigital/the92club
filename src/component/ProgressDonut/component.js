import React, { Component } from 'react'
import PropTypes from 'prop-types'

import {
  Container,
  InfoContainer,
  Info,
  AchievedCount,
  GoalCount,
  GoalOf,
  GoalTitle,
} from './styled'

function polarToCartesian(centerX, centerY, radius, angleInDegrees) {
  const angleInRadians = (angleInDegrees - 90) * (Math.PI / 180.0)
  return {
    x: centerX + radius * Math.cos(angleInRadians),
    y: centerY + radius * Math.sin(angleInRadians),
  }
}

function describeArc(x, y, radius, startAngle, endAngle) {
  const start = polarToCartesian(x, y, radius, endAngle)
  const end = polarToCartesian(x, y, radius, startAngle)
  const largeArcFlag = endAngle - startAngle <= 180 ? '0' : '1'
  const d = ['M', start.x, start.y, 'A', radius, radius, 0, largeArcFlag, 0, end.x, end.y].join(' ')
  return d
}

let timer
const timerDelayMilliseconds = 10

function getPercentageAchieved({ goal, achieved }) {
  const percent = Math.round(achieved / goal * 100)
  if (percent < 0) {
    return 0
  }
  if (percent > 100) {
    return 100
  }
  return percent
}

class ProgressDonut extends Component {
  constructor(props) {
    super(props)
    this.state = {
      percentageFilteredTarget: getPercentageAchieved({
        achieved: props.achieved,
        goal: props.goal,
      }),
      percentageFiltered: 0,
      achievedDisplayTarget: props.achieved,
      achievedDisplay: 0,
    }
  }

  componentWillMount() {
    this.animateState()
  }

  componentWillReceiveProps(nextProps) {
    if (nextProps.goal !== this.props.goal || nextProps.achieved !== this.props.achieved) {
      const target = getPercentageAchieved({
        goal: nextProps.goal,
        achieved: nextProps.achieved,
      })
      this.setState({
        percentageFilteredTarget: Number(target),
        achievedDisplayTarget: Number(nextProps.achieved),
      })
    }
  }

  componentDidUpdate() {
    this.animateState()
  }

  animateState() {
    const {
      percentageFiltered: percentageFilteredState,
      percentageFilteredTarget: percentageFilteredTargetState,
      achievedDisplay: achievedDisplayState,
      achievedDisplayTarget: achievedDisplayTargetState,
    } = this.state
    if (
      percentageFilteredState !== percentageFilteredTargetState ||
      achievedDisplayState !== achievedDisplayTargetState
    ) {
      clearInterval(timer)
      timer = setInterval(() => {
        this.setState((prevState) => {
          const {
            percentageFiltered,
            percentageFilteredTarget,
            achievedDisplay,
            achievedDisplayTarget,
          } = prevState
          let newPercentageFiltered = percentageFiltered
          let newAchievedDisplay = achievedDisplay
          if (percentageFiltered !== percentageFilteredTarget) {
            const incrementDecrementValue = percentageFilteredTarget > percentageFiltered ? 1 : -1
            newPercentageFiltered = percentageFiltered + incrementDecrementValue
          }
          if (achievedDisplay !== achievedDisplayTarget) {
            const diff = Math.abs(achievedDisplayTarget - achievedDisplay)
            const percentDiff = Math.abs(percentageFilteredTarget - percentageFiltered)
            const incrementalDiff = Math.round(diff / percentDiff)
            const incrementDecrementValue = achievedDisplayTarget > achievedDisplay
              ? incrementalDiff
              : -incrementalDiff
            newAchievedDisplay = achievedDisplay + incrementDecrementValue
            if (isNaN(newAchievedDisplay) || !isFinite(newAchievedDisplay)) {
              newAchievedDisplay = achievedDisplayTarget
            }
          }
          return {
            percentageFiltered: newPercentageFiltered,
            achievedDisplay: newAchievedDisplay,
          }
        })
      }, timerDelayMilliseconds)
    } else {
      clearInterval(timer)
    }
  }

  render() {
    const { baseTitle, goal, width, height, baseColor, resultsColor, strokeWidth } = this.props
    const { percentageFiltered, achievedDisplay } = this.state
    const base = describeArc(150, 150, 100, 0, 360 - 0.0001)
    const arcDegrees = percentageFiltered === 0 ? 360 - 0.0001 : 360 * (percentageFiltered / 100)
    let arc
    if (arcDegrees === 360) {
      arc = describeArc(150, 150, 100, 0, 360 - 0.0001)
    } else {
      arc = describeArc(150, 150, 100, 0, arcDegrees)
    }
    const widgetStyles = {
      width,
      height,
      position: 'absolute',
      display: 'flex',
      justifyContent: 'center',
      alignItems: 'center',
    }

    return (
      <Container width={width} height={height}>
        <svg className="results-donut" viewBox="42 42 216 216" style={widgetStyles}>
          <path d={base} fill="none" stroke={baseColor} strokeWidth={strokeWidth} />
        </svg>
        {percentageFiltered > 0 &&
          <svg className="results-donut" viewBox="42 42 216 216" style={widgetStyles}>
            <path d={arc} fill="none" stroke={resultsColor} strokeWidth={strokeWidth} />
          </svg>}
        <InfoContainer width={width} height={height}>
          <Info>
            <AchievedCount>{achievedDisplay}</AchievedCount>
            <GoalOf>of</GoalOf>
            <GoalCount>{goal}</GoalCount>
            <GoalTitle>{baseTitle}</GoalTitle>
          </Info>
        </InfoContainer>
      </Container>
    )
  }
}

ProgressDonut.displayName = 'ProgressDonut'

ProgressDonut.propTypes = {
  baseTitle: PropTypes.string.isRequired,
  goal: PropTypes.number.isRequired,
  achieved: PropTypes.number.isRequired,
  strokeWidth: PropTypes.number,
  width: PropTypes.number,
  height: PropTypes.number,
  baseColor: PropTypes.string,
  resultsColor: PropTypes.string,
}

ProgressDonut.defaultProps = {
  strokeWidth: 15,
  width: 150,
  height: 150,
  resultsColor: '#01E400',
  baseColor: '#BBB',
}

export default ProgressDonut
