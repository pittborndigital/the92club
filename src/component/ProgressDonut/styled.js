import styled from 'styled-components'

export const Container = styled.div`
  position: relative;
  height: ${props => `${props.height}px`};
  width: ${props => `${props.width}px`};
`
Container.displayName = 'ProgressDonut.Container'

export const InfoContainer = styled.div`
  position: absolute;
  display: flex;
  justify-content: center;
  align-items: center;
  height: ${props => `${props.height}px`};
  width: ${props => `${props.width}px`};
`
InfoContainer.displayName = 'ProgressDonut.InfoContainer'

export const Info = styled.div`
  text-align: center;
`
Info.displayName = 'ProgressDonut.Info'

export const AchievedCount = styled.div`
  font-size: 26px;
  font-weight: bold;
`
AchievedCount.displayName = 'ProgressDonut.AchievedCount'

export const GoalCount = styled.div`
  font-size: 16px;
  padding: 2px;

`
GoalCount.displayName = 'ProgressDonut.GoalCount'

export const GoalOf = styled.div`
  font-size: 12px;
`
GoalOf.displayName = 'ProgressDonut.GoalOf'

export const GoalTitle = styled.div`
  font-size: 12px;
  padding-bottom: 6px;
`
GoalTitle.displayName = 'ProgressDonut.GoalTitle'
