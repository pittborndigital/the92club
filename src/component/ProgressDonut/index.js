import ProgressDonut from './component'

export default ProgressDonut
