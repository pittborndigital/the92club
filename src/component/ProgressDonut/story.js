import React from 'react'

import { storiesOf } from '@storybook/react'
import { withKnobs, number, text } from '@storybook/addon-knobs'

import ProgressDonut from './component'

const stories = storiesOf('ProgressDonut', module)
stories.addDecorator(withKnobs)

stories.add('default', () =>
  (<ProgressDonut
    baseTitle={text('Base Title', 'Grounds')}
    goal={number('Goal', 92)}
    achieved={number('Achieved', 83)}
  />),
)

stories.add('all options', () =>
  (<ProgressDonut
    baseTitle={text('Base Title', 'Grounds')}
    goal={number('Goal', 92)}
    achieved={number('Achieved', 83)}
    strokeWidth={number('strokeWidth', 15)}
    width={number('width', 150)}
    height={number('height', 150)}
    resultsColor={text('resultsColor', 'green')}
    baseColor={text('baseColor', 'gray')}
  />),
)

stories.add('complete', () => <ProgressDonut baseTitle="Grounds" goal={92} achieved={92} />)
