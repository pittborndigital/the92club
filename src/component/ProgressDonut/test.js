import React from 'react'
import renderer from 'react-test-renderer'

import ProgressDonut from './component'

it('renders correctly', () => {
  const tree = renderer
    .create(<ProgressDonut goal={38} achieved={64} goalTitle="Donuts Eaten" />)
    .toJSON()
  expect(tree).toMatchSnapshot()
})
