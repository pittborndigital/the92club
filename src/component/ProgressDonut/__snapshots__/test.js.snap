// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="sc-bdVaJa kNeXYG"
  height={150}
  width={150}
>
  <svg
    className="results-donut"
    style={
      Object {
        "alignItems": "center",
        "display": "flex",
        "height": 150,
        "justifyContent": "center",
        "position": "absolute",
        "width": 150,
      }
    }
    viewBox="42 42 216 216"
  >
    <path
      d="M 149.9998254670748 50.00000000015231 A 100 100 0 1 0 150 50"
      fill="none"
      stroke="#BBB"
      strokeWidth={15}
    />
  </svg>
  <div
    className="sc-bwzfXH gBPZuy"
    height={150}
    width={150}
  >
    <div
      className="sc-htpNat dEQUjC"
    >
      <div
        className="sc-bxivhb kpjUiq"
      >
        0
      </div>
      <div
        className="sc-EHOje hUWbrG"
      >
        of
      </div>
      <div
        className="sc-ifAKCX cjzFD"
      >
        38
      </div>
      <div
        className="sc-bZQynM bQGae"
      >
        Donuts Eaten
      </div>
    </div>
  </div>
</div>
`;
